nixos/modules/services/matrix/conduwuit.nix

```diff
@@ -0,0 +1,280 @@
+              is set to true, users can freely register.
+            '';
+          };
+          global.allow_encryption = lib.mkOption {
+            type = lib.types.bool;
+            default = true;
+            description = "Whether new encrypted rooms can be created. Note: existing rooms will continue to work.";
+          };
+          global.allow_federation = lib.mkOption {
+            type = lib.types.bool;
+            default = true;
+            description = ''
+              Whether this server federates with other servers.
+            '';
+          };
+          global.trusted_servers = lib.mkOption {
+            type = lib.types.listOf lib.types.nonEmptyStr;
+            default = [ "matrix.org" ];
+            description = ''
+              Servers listed here will be used to gather public keys of other servers
+              (notary trusted key servers).
+
+              Currently, conduwuit doesn't support inbound batched key requests, so
+              this list should only contain other Synapse servers.
+
+              Example: `[ "matrix.org" "constellatory.net" "tchncs.de" ]`
+            '';
+          };
+          global.database_path = lib.mkOption {
+            readOnly = true;
+            type = lib.types.path;
+            default = "/var/lib/conduwuit/";
+            description = ''
+              Path to the conduwuit database, the directory where conduwuit will save its data.
+              Note that database_path cannot be edited because of the service's reliance on systemd StateDir.
+            '';
+          };
+          global.allow_check_for_updates = lib.mkOption {
+            type = lib.types.bool;
+            default = false;
+            description = ''
+              If enabled, conduwuit will send a simple GET request periodically to
+              <https://pupbrain.dev/check-for-updates/stable> for any new announcements made.
+              Despite the name, this is not an update check endpoint, it is simply an announcement check endpoint.
+
+              Disabled by default.
+            '';
+          };
+        };
+      };
+      default = { };
+      description = ''
+        Generates the conduwuit.toml configuration file. Refer to
+        <https://conduwuit.puppyirl.gay/configuration.html>
+        for details on supported values.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = cfg.settings.global.unix_socket_path == null || cfg.settings.global.address == null;
+        message = ''
+          In `services.conduwuit.settings.global`, `unix_socket_path` and `address` cannot be set at the
+          same time.
+          Leave one of the two options unset or explicitly set them to `null`.
+        '';
+      }
+    ];
+
+    # DynamicUser requires both user and group to exist or both not to exist.
+    # If we're configured to run under a different user or group, ensure
+    # the other default exists.
```

Comment:
This would be much better placed in assertion, as this requirement is easily broken with the following

```nix
import ./nixos {
  configuration = {
    services.conduwuit = {
      enable = true;

      user = "testUser";
    };
  };
}
```

```
nix-repl> conduwuitUser = config.services.conduwuit.user

nix-repl> conduwuitGroup = config.services.conduwuit.group

nix-repl> config.users.users.${conduwuitUser}
error: attribute 'testUser' missing
       at «string»:1:1:
            1| config.users.users.${conduwuitUser}
             | ^

nix-repl> config.users.groups.${conduwuitGroup}
evaluation warning: system.stateVersion is not set, defaulting to 25.05. Read why this matters on https://nixos.org/manual/nixos/stable/options.html#opt-system.stateVersion.
{
  gid = null;
  members = [ ... ];
  name = "conduwuit";
}
```

Reply:
I don't see how this is different from how any other NixOS module works that allows changing the user. The user is only ever created for all NixOS modules I checked if it's set to the default value.

Reply:
But to make it somewhat less confusing, I now always create the user if its set to the default user, and the group if its set to the default group, as this does not conflict with DynamicUser (which will remain set as it enables some more security options in systemd).

Reply:
I've also added assertions now to ensure the configured user or group exist if they are changed.